### PR TITLE
Rework the notification system

### PR DESCRIPTION
**Pull request by almet**
The current notification system has some issues:

- Notifications stay until you dismiss them, it should be an option,
- When you dismiss them, you're being redirected, whereas that's not the behaviour we would like,
- We should be able to attach some actions inside the notifications (for instance, undo), but we currently don't have any way to do this.

### CODE REVIEW

**Comment by Natim** [comment]
The linter said: " 8:3  error  "NOTIFICATION_REMOVE" is defined but never used  no-unused-vars"

**Comment by almet** [comment]
I've updated the PR, it had an un-used import.

File: formbuilder/actions/notifications.js [hunk 1]
```diff
@@ -1,10 +1,33 @@
...
+    type: "info",
+    autoDismiss: null,
+    dismissAfter: 5000
+  };
+  let {type, autoDismiss, dismissAfter} = {
+    ...defaultOptions,
+    ...options
+  };
+  if (autoDismiss == null) {
+    autoDismiss = (type == "error") ? false : true;
```
Comment (n1k0) [inline formbuilder/actions/notifications.js]: Nit: `autoDismiss = type !== "error"`

File: formbuilder/actions/notifications.js [hunk 2]
```diff
@@ -1,10 +1,33 @@
...
+  }
+  return (dispatch) => {
+    const id = uuid.v4();
+    dispatch({type: NOTIFICATION_ADD, notification: {id, message, type}});
+    if (autoDismiss == true) {
```
Comment (n1k0) [inline formbuilder/actions/notifications.js]: Nit: ===

File: formbuilder/components/NotificationList.js [hunk 3]
```diff
@@ -12,13 +12,13 @@ export default function NotificationList(props) {
...
+      notifications.map((notif) => {
+        const {message, type, id} = notif;
         const classes = `alert alert-${ERROR_CLASSES[type]}`;
         return (
-          <div key={index} className={classes}>
-            <a href="#" className="close"
-              onClick={() => removeNotification(index)}>×</a>
+          <div key={id} className={classes}>
+            <a className="close"
+               onClick={() => removeNotification(id)}>×</a>
```
Comment (n1k0) [inline formbuilder/components/NotificationList.js]: Nit: 2 spaces indent was fine

File: formbuilder/reducers/notifications.js [hunk 4]
```diff
@@ -12,7 +12,7 @@ export default function collections(state = INITIAL_STATE, action) {
     return [...state, action.notification];
 
   case NOTIFICATION_REMOVE:
-    return state.filter((notif, index) => action.index !== index);
+    return state.filter((obj) => action.id !== obj.id);
```
Comment (n1k0) [inline formbuilder/reducers/notifications.js]: Nit: 

```js
return state.filter(({id}) => action.id !== id);
```

File: test/actions/notifications_test.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  describe("addNotification", () => {
+    it("should send info notifications by default", (done) => {
+      addNotification("Some info message")(({type, notification}) => {
+        expect(type).to.eql(NOTIFICATION_ADD);
+        expect(notification.message ).to.eql("Some info message");
```
Comment (n1k0) [inline test/actions/notifications_test.js]: Nit: unwanted space after .message

File: test/actions/notifications_test.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        expect(notification.type).to.eql("info");
+        done();
+      });
+    });
+
+    it("should send error notifications if specified", (done) => {
+      addNotification("Some error message", {type: "error"})(({type, notification}) => {
```
Comment (n1k0) [inline test/actions/notifications_test.js]: Nit: maybe slightly more readable with an alternative const for the thunk:

```js
const thunk = addNotification("Some error message", {type: "error"});
thunk(({type, notification}) => {
  ...
```

File: test/actions/server_test.js [hunk 7]
```diff
@@ -23,6 +23,23 @@ import {
   NOTIFICATION_ADD
 } from "../../formbuilder/actions/notifications";
 
+function getErrorDispatch(done) {
+  return (func) => {
+    if (typeof func === "function") {
```
Comment (n1k0) [inline test/actions/server_test.js]: Nit: returning early when `func` is not a function might ease legibility a little

**Comment by n1k0** [comment]
Didn't test the patch live the code looks good!

**Comment by almet** [comment]
Thanks for the review, I've pushed some changes after your review. Let's merge!

**Pull request merged**